api/pkg/client/mlflow/mlflow.go

```diff
@@ -0,0 +1,173 @@
+	}
+
+	return nil
+}
+
+func (mfc *mlflowClient) SearchRunForExperiment(idExperiment string) (SearchRunsResponse, error) {
+	// Search related runs for an experiment id
+	var responseObject SearchRunsResponse
+
+	searchRunURL := fmt.Sprintf("%s/api/2.0/mlflow/runs/search", mfc.Config.TrackingURL)
+
+	input := SearchRunRequest{ExperimentId: []string{idExperiment}}
+	jsonInput, err := json.Marshal(input)
+	if err != nil {
+		return responseObject, err
+	}
+
+	err = mfc.httpCall("POST", searchRunURL, jsonInput, &responseObject)
+	if err != nil {
+		return responseObject, err
+	}
+
+	return responseObject, nil
+}
+
+func (mfc *mlflowClient) SearchRunData(idRun string) (SearchRunResponse, error) {
+	// Creating Output Format for Run Detail
+	var runResponse SearchRunResponse
+	getRunURL := fmt.Sprintf("%s/api/2.0/mlflow/runs/get?run_id=%s", mfc.Config.TrackingURL, idRun)
+
+	err := mfc.httpCall("GET", getRunURL, nil, &runResponse)
+	if err != nil {
+		return runResponse, err
+	}
+	return runResponse, nil
+}
+
+func (mfc *mlflowClient) DeleteExperiment(idExperiment string) error {
+
+	relatedRunId, err := mfc.SearchRunForExperiment(idExperiment)
+	if err != nil {
+		return err
+	}
+	// Error Handling, keep runId that failed to delete
+	// TODO: What should i do with this?
+	var deletedRunId []string
+	var failDeletedRunId []string
+	for _, run := range relatedRunId.RunsData {
+		err = mfc.DeleteRun(run.Info.RunId, false)
+		if err != nil {
+			failDeletedRunId = append(failDeletedRunId, run.Info.RunId)
+			// return err
+		} else {
+			deletedRunId = append(deletedRunId, run.Info.RunId)
+		}
+	}
+
+	if len(relatedRunId.RunsData) > 0 {
+		// the [5:] is to remove the "gs://" on the artifact uri
+		// ex : gs://bucketName/path → bucketName/path
+		path := relatedRunId.RunsData[0].Info.ArtifactURI[5:]
+
+		// This section is getting gcs path for a folder, from a run id
+		// ex : bucketName/mlflow/idExperiment/runId/artefact  → bucketName/path/mlflow/idExperiment
+		splitPath := strings.SplitN(path, "/", 4)
+		folderPath := strings.Join(splitPath[0:3], "/")
+		// deleting folder
+		err = mfc.GcsService.DeleteArtifact(folderPath)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (mfc *mlflowClient) DeleteRun(idRun string, deleteArtifact bool) error {
+	// Creating Input Format for Delete run
+	input := DeleteRunRequest{RunId: idRun}
+	// HIT Delete Run API
+	delRunURL := fmt.Sprintf("%s/api/2.0/mlflow/runs/delete", mfc.Config.TrackingURL)
+
+	jsonInput, err := json.Marshal(input)
+	if err != nil {
+		return err
+	}
+
+	err = mfc.httpCall("POST", delRunURL, jsonInput, nil)
+	if err != nil {
+		return err
+	}
+
+	if deleteArtifact {
+		runDetail, err := mfc.SearchRunData(idRun)
+		if err != nil {
+			return err
+		}
+		// the [5:] is to remove the "gs://" on the artifact uri
+		// ex : gs://bucketName/path → bucketName/path
+		err = mfc.GcsService.DeleteArtifact(runDetail.RunData.Info.ArtifactURI[5:])
+		if err != nil {
+			return err
+		}
```

Comment:
On the topic of error handling, I was also wondering about the steps that should be taken if any of these two operations fail, since the runs would've already been deleted from the mlflow database. Should the runs be restored in these cases, i.e. 

```
delete experiment/run in mlflow database
if err
    return err
delete artefact in GCS
if err
    restore experiment/run in mlflow database
    return err
```

That would allow a user to attempt deleting the GCS artefact again once the issue has been resolved, as opposed to having a GCS artefact left hanging once an error has been thrown since the run would no longer be accessible by the user (it has already been deleted from the mlflow database). The only way to delete the GCS artefact then would be to remove it manually in this case.

@krithika369 what do you think? @Zenovore and I talked about this separately and he thinks this is feasible

Reply:
Oh.. in this case, can we delete the GCS Artifact first before soft deleting the run? The workflow will be something like:

* Call Merlin's delete model
* Call MLP Delete Experiment: For each "active" (not already soft-deleted) Run in Experiment:
    - Call MLP Delete Run: This will delete the GCS artifact first, if exists ("if exists" because a previous failed deletion attempt may have been made which did some partial clean up). Then, it will call MLFLow Run delete (soft delete).
* If all operations above succeed, we proceed to delete the Merlin model data from the DB. If any intermediate step fails, we should stop the whole flow.

Does it make sense?

Reply:
This way, we don't need to restore the run on error:

> if err
    restore experiment/run in mlflow database

Reply:
> Call MLP Delete Experiment: For each "active" (not already soft-deleted) Run in Experiment:
Call MLP Delete Run: This will delete the GCS artifact first, if exists ("if exists" because a previous failed deletion attempt may have been made which did some partial clean up). Then, it will call MLFLow Run delete (soft delete).

Understood, i'll switch the flow to deleting the GCS artifact first, and then soft-deleting the run, this approach doesn't need restoring any runs.